test/python/test_apps.py

```diff
@@ -57,12 +65,156 @@ def test_trial_functions(self):
         self.log.info(qasm_txt)
         self.assertEqual(len(qasm_txt), 564)
 
+    def test_optimization_of_H2_at_bond_length(self):
+        """ From https://github.com/QISKit/qiskit-tutorial/blob/master/\
+        4_applications/quantum_chemistry.ipynb#Optimization-of-H2-at-bond-length but shorter."""
+        n = 2
+        m = 6
+        device = 'local_qasm_simulator'
+
+        np.random.seed(42)
+        initial_theta = np.random.randn(2 * n * m)
+        entangler_map = {
+            0: [1]}  # the map of two-qubit gates with control at key and target at values
+        shots = 1
+        max_trials = 1
+        ham_name = self._get_resource_path("../performance/H2/H2Equilibrium.txt")
+
+        # Exact Energy
+        pauli_list = Hamiltonian_from_file(ham_name)
+        H = make_Hamiltonian(pauli_list)
+        exact = np.amin(la.eig(H)[0]).real
+        self.log.info('The exact ground state energy is: %s', exact)
+        self.assertEqual(exact, -1.8572746704950902)
+
+        # Optimization
+        Q_program = QuantumProgram()
+
+        def cost_function(Q_program, H, n, m, entangler_map, shots, device, theta):
+            # pylint: disable=missing-docstring
+            return eval_hamiltonian(Q_program, H,
+                                    trial_circuit_ryrz(n, m, theta, entangler_map, None, False),
+                                    shots, device).real
+
+        initial_c = 0.01
+        target_update = 2 * np.pi * 0.1
+        save_step = 20
```

Comment:
Warning: save_step = 20 doesn't make sense if max_trials is 1

Reply:
I would set max_trials > save_step if we want to check partial optimization info messages

Reply:
Since `save_steps` and `last_avg` are optional, I will just ignore them. (see 2ad3af53b186c0cb561c1da1be79a952de2ed6d0)